File: top-k-frequent-elements/topkFrequentElement.java
public class Solution {
    //O(n) priority queue
    class Pair {
        int value;
        int count;
        
        public Pair(int value, int count) {
            this.value = value;
            this.count = count;
        }
    }
    public List<Integer> topKFrequent(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return null;
        }
        PriorityQueue<Pair> pq = new PriorityQueue<>(new Comparator<Pair>() {
            public int compare(Pair a, Pair b) {
                return b.count - a.count;
            }
        });
        
        List<Integer> res = new ArrayList<>();
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int i : nums) {
            if (!map.containsKey(i)) {
                map.put(i, 1);
            } else {
                int count = map.get(i);
                map.put(i, count + 1);
            }
        }
        
        for (Integer i : map.keySet()) {
            int count = map.get(i);
            Pair p = new Pair(i, count);
            pq.add(p);
        }
        for (int i = 0; i < k; ++i) {
            res.add(pq.poll().value);
        }
        return res;
    }

    //O(n) buckets
    public List<Integer> topKFrequent(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return null;
        }
        
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        
        List<Integer>[] buckets = new List[nums.length + 1];
        for (Integer i : map.keySet()) {
            int count = map.get(i);
            if (buckets[count] == null) {
                buckets[count] = new ArrayList<>();
            }
            buckets[count].add(i);
        }
        
        int size = k;
        List<Integer> res = new ArrayList<>();
        for (int i = buckets.length - 1; i > 0 && size > 0; --i) {
            if (buckets[i] != null) {
                res.addAll(buckets[i]);
                size -= buckets[i].size();
            }
        }
        return res;
    }
}

File: top-k-frequent-elements/topkFrequentElement.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        if nums == None or len(nums) == 0:
            return None
        
        hashmap = {}
        for i in nums:
            if i in hashmap:
                hashmap[i] += 1
            else:
                hashmap[i] = 1
        
        res = []
        buckets = [[] for i in range(len(nums) + 1)]
        for i in hashmap.keys():
            count = hashmap[i]
            buckets[count].append(i)
        
        for i in range(len(buckets) - 1, 0, -1):
            if len(buckets[i]) != 0:
                res += buckets[i]
            if len(res) >= k:
                break;
        
        return res